### PR TITLE
feat(platform/builder): Builder credentials support + UX improvements

### LINKED ISSUE #10313: Fix graph run mechanism in builder
* Pass in graph inputs to the execute call rather than saving them as defaults in the graph
* Add graph-level credentials input support

### LINKED ISSUE #10333: Diagnose & fix Builder sluggishness
Moving the viewport around is not smooth at all, probably because the app is re-rendering stuff it shouldn't. Fixing this would be a big UX polish improvement.

### PR DESCRIPTION
**Pull request by Pwuts**
- Resolves #10313
- Resolves #10333

Before:

https://github.com/user-attachments/assets/a105b2b0-a90b-4bc6-89da-bef3f5a5fa1f
- No credentials input
- Stuttery experience when panning or zooming the viewport

After:

https://github.com/user-attachments/assets/f58d7864-055f-4e1c-a221-57154467c3aa
- Pretty much the same UX as in the Library, with fully-fledged credentials input support
- Much smoother when moving around the canvas

### Changes 🏗️

Frontend:
- Add credentials input support to Run UX in Builder
  - Pass run inputs instead of storing them on the input nodes
  - Re-implement `RunnerInputUI` using `AgentRunDetailsView`; rename to `RunnerInputDialog`
    - Make `AgentRunDraftView` more flexible
    - Remove `RunnerInputList`, `RunnerInputBlock`
- Make moving around in the Builder *smooooth* by reducing unnecessary re-renders
  - Clean up and partially re-write bead management logic
- Replace `request*` fire-and-forget methods in `useAgentGraph` with direct action async callbacks
- Clean up run input UI components
  - Simplify `RunnerUIWrapper`
  - Add `isEmpty` utility function in `@/lib/utils` (expanding on `_.isEmpty`)
  - Fix default value handling in `TypeBasedInput` (**Note:** after all the changes I've made I'm not sure this is still necessary)
- Improve & clean up Builder test implementations

Backend + API:
- Fix front-end `Node`, `GraphMeta`, and `Block` types
- Small refactor of `Graph` to match naming of some `LibraryAgent` attributes
- Fix typing of `list_graphs`, `get_graph_meta_by_store_listing_version_id` endpoints
  - Add `GraphMeta` model and `GraphModel.meta()` shortcut
- Move `POST /library/agents/{library_agent_id}/setup-trigger` to `POST /library/presets/setup-trigger`

#### For code changes:
- [x] I have clearly listed my changes in the PR description
- [x] I have made a test plan
- [x] I have tested my changes according to the test plan:
  - Test the new functionality in the Builder:
    - [x] Running an agent with (credentials) inputs from the builder
      - [x] Beads behave correctly
    - [x] Running an agent without any inputs from the builder
    - [x] Scheduling an agent from the builder
    - [x] Adding and searching blocks in the block menu
  - [x] Test that all existing `AgentRunDraftView` functionality in the Library still works the same
    - [x] Run an agent
    - [x] Schedule an agent
    - [x] View past runs
  - [x] Run an agent with inputs, then edit the agent's inputs and view the agent in the Library (should be fine)

### CODE REVIEW

**Comment by AutoGPT-Agent** [comment]
Thanks for this PR improving the graph run mechanism! The code changes look promising, making good architectural improvements by passing inputs directly instead of storing them on input nodes.

However, before we can merge this, please complete the checklist in the PR description:

- The most important item missing is your test plan - what scenarios did you test to ensure these changes work correctly?
- Please confirm you've tested the functionality for both regular graph runs and scheduled runs

Also, could you provide a bit more detail about the architectural improvements you're making? The current description mentions "Pass run inputs instead of storing them on the input nodes" but it would be helpful to understand:

1. Why this approach is better
2. What specific issues it solves from the referenced issue #10313

This will help reviewers understand the context and intent of the changes better. The code changes themselves look well-structured, but we need the checklist completed before merging.

**Comment by AutoGPT-Agent** [comment]
Thank you for your PR to improve the graph run mechanism! The changes look promising, cleaning up the run input UI components and modifying how run inputs are passed instead of storing them on input nodes.

However, before we can approve this PR, please complete the checklist in the PR description. Currently, none of the items are checked off. Please mark the items as completed or remove irrelevant sections if they don't apply.

Specifically:
- Check the "I have clearly listed my changes in the PR description" item
- Complete your test plan and check that item
- Check that you've tested your changes according to the test plan

Also, I noticed the PR has a "conflicts" label - please resolve any merge conflicts before we can proceed with the review.

The code changes themselves look good - the refactoring to pass run inputs directly instead of storing them on input nodes is a good improvement to the architecture.

**Comment by AutoGPT-Agent** [comment]
Thank you for your PR to improve the graph run mechanism in the frontend builder! The code changes look well-structured and focused on the task of refactoring how run inputs are handled.

However, I notice that the checklist in your PR description hasn't been completed. Before we can merge this PR, could you please:

1. Check off the items in the checklist that you have completed
2. Add a test plan that describes how you've tested these changes
3. Complete any remaining tests according to your test plan

Your changes appear to be significant, involving multiple components related to the run input mechanism. A thorough test plan would help reviewers understand how you've verified that these changes work correctly in different scenarios.

Once you've updated the PR description with a completed checklist and test plan, we'll be able to proceed with the review process.

**Comment by AutoGPT-Agent** [comment]
Thank you for your PR to improve the graph run mechanism in the frontend builder. The changes look well-structured and address the issue mentioned in #10313.

However, I noticed that the checklist in your PR description is not completely filled out. Our guidelines require all checklist items to be checked off before a PR can be merged. Specifically, please complete these items:

- [x] I have clearly listed my changes in the PR description
- [x] I have made a test plan
- [x] I have tested my changes according to the test plan

Please also add your test plan in the appropriate section so we can understand how you've verified these changes.

The actual code changes look good - you've improved how run inputs are handled by passing them directly instead of storing them on input nodes, which should make the system more maintainable. I particularly appreciate the clean refactoring of the RunnerUIWrapper and other UI components.

Once you've completed the checklist with your test plan, this PR should be ready for another review.

**Comment by AutoGPT-Agent** [comment]
Thank you for this PR to improve the graph run mechanism!

The changes look good overall - I can see you've refactored how run inputs are handled, moving from storing them on input nodes to passing them directly. The code is well-structured and the refactoring seems comprehensive across both frontend and backend components.

However, before this can be merged:

### Required Fixes

- **Complete the checklist in the PR description**: Please check off all the boxes in the checklist and add your test plan. This is required for all PRs with code changes.

### Additional Comments

- I like the approach to clean up the run input UI components and the refactoring to pass run inputs instead of storing them on nodes
- The backend changes to support this look appropriate and maintain proper security patterns
- The new GraphMeta class is a nice improvement for returning metadata without exposing node/link details

Once you've completed the checklist with a proper test plan, this PR should be ready for another review.

**Comment by AutoGPT-Agent** [comment]
Thank you for your contribution to improve the graph run mechanism. I've reviewed your PR and noticed a few issues that need to be addressed before it can be merged:

1. **Incomplete Checklist**: The checklist in your PR description is not filled out. Please either:
   - Complete the checklist by checking off all items
   - If these changes are non-material (refactoring only), you can remove the checklist section entirely

2. **Test Plan**: Since this appears to be a significant refactoring of the run input handling mechanism, please include a test plan describing how you've verified the changes work correctly.

Your changes look well-structured and focused on the stated goal of improving the graph run mechanism. The PR does a good job of consolidating run input handling and cleaning up the UI components related to this functionality.

Once you've addressed the checklist requirements, this PR should be ready for another review.

**Comment by AutoGPT-Agent** [comment]
Thank you for your work on improving the graph run mechanism! Before we can merge this PR, there are a few process items that need to be addressed:

1. **PR Description**: The description starts with changes but cuts off with "...". Please complete the description to fully reflect all changes in this PR, which is quite extensive. A thorough description helps reviewers understand the scope and intent of your changes.

2. **Checklist**: The checklist in the PR description is completely unchecked. Please fill out the checklist items, particularly focusing on:
   - Clearly listing all changes
   - Including a test plan
   - Documenting how you've tested these changes

The code changes themselves look good - I appreciate the comprehensive refactoring of the run mechanism, making it more maintainable and consistent. The separation of concerns between different components is much improved.

Once you've updated the description and completed the checklist, we'll be able to move forward with the review process.

**Comment by AutoGPT-Agent** [comment]
## PR Review Feedback

Thank you for your work on improving the graph run mechanism. This PR introduces significant improvements to how graph runs are handled. However, there are a few issues that need to be addressed before this can be merged:

### Required Changes

1. **Incomplete Checklist**: The PR checklist is completely unchecked. Please fill out the checklist completely, including your test plan. This is a mandatory requirement for all PRs.

2. **Scope Mismatch**: While your PR title indicates this is a frontend/builder change, the PR contains substantial backend changes as well, including:
   - Significant modifications to `backend/data/graph.py`
   - Adding a new `GraphMeta` class
   - Renaming functions like `get_graphs` to `list_graphs`
   - Moving endpoint functionality

   Please consider either:
   - Updating the PR title to better reflect the full scope of changes (e.g., `feat(platform): Improve graph run mechanism`), or
   - Splitting this into separate PRs for frontend and backend changes

### Positive Aspects

- The refactoring of run input handling appears to be a significant improvement
- The code looks well-structured and follows consistent patterns
- The changes maintain proper user_id security checks
- The reorganization of the API appears to make the codebase more maintainable

Please address the checklist issue at minimum before this PR can be merged. Let me know if you need any clarification or assistance with these changes.

File: autogpt_platform/frontend/src/components/type-based-input.tsx
Comment (Pwuts) [inline autogpt_platform/frontend/src/components/type-based-input.tsx:1]: **Note:** after replacing `RunnerInputList` with `AgentRunDraftView` I'm not sure these changes are still necessary

**Comment by AutoGPT-Agent** [comment]
Thank you for your detailed PR! The code changes look well-structured and focused on improving the graph run mechanism.

However, before we can approve this PR, we need a completed test plan in your checklist. Given the scope of your changes (including backend API modifications and frontend UI refactoring), it's important to document how you've tested these changes.

Please update your PR description to include:

- A completed test plan section in the checklist detailing what scenarios you've tested
- Check off the remaining checkboxes in the PR description

Some testing scenarios to consider:
- Testing with credentials inputs
- Testing run inputs with and without saved presets
- Testing the new edge/bead rendering optimizations
- Verifying API compatibility with the endpoint changes

Once you've added this information, we'll be happy to review again!

**Comment by AutoGPT-Agent** [comment]
Thanks for your detailed PR submission to improve the graph run mechanism! Before we can approve this, there are a couple of items that need to be addressed:

1. **Incomplete checklist:** You've checked off that you've listed your changes, but the test plan section is incomplete. Please fill out a test plan detailing how you've verified the changes work as expected.

2. **Scope mismatch:** Your PR title indicates this is a frontend/builder feature, but there are significant backend changes as well (roughly 30-35% of the diff). These include:
   - Adding a new `GraphMeta` model
   - Renaming API endpoints
   - Moving webhook trigger setup functionality from library agents to presets
   - Changes to various backend data models

Please either:
   - Update your PR title to reflect both frontend and backend changes (e.g., `feat(platform): Improve graph run mechanism`) 
   - OR split this PR into separate frontend and backend PRs if that makes more sense

The code changes themselves look well-structured and the refactoring seems beneficial, particularly how you've reduced unnecessary re-renders and improved the run input handling. Once you've addressed the checklist and scope issues, we should be ready to proceed with the review.

**Comment by AutoGPT-Agent** [comment]
Thank you for your comprehensive PR! The changes to improve the graph run mechanism look well-structured and thoughtful.

I noticed a few items that need to be addressed before this can be merged:

1. **Incomplete Checklist**: Your PR checklist is partially completed. Please fill out the test plan section or mark it as N/A if not applicable.

2. **TODO Item**: You mention a "weird double-close bug on `RunnerInputDialog`" in your TODO list. Is this something that should be fixed before merging, or can it be addressed in a follow-up PR?

The code changes look solid with good organization and improved naming. I especially appreciate:
- Moving from fire-and-forget methods to direct async callbacks in `useAgentGraph`
- Reducing unnecessary re-renders in the Builder
- The refactoring of `Graph` to match naming conventions with `LibraryAgent`

Once you complete the checklist with a test plan, this should be ready for review.

**Comment by AutoGPT-Agent** [comment]
Thanks for this comprehensive PR improving the Builder UX and adding credentials support! I appreciate the detailed description of your changes.

## Issues to Address

### Required Test Plan
- The PR checklist indicates you haven't created a test plan yet. Please fill out the test plan section with specific steps to verify your changes.
- Given the scope of changes (frontend Builder UX + credentials support), your test plan should cover:
  - Testing the new credentials input support in the Builder
  - Verifying the improved movement/performance in the Builder
  - Testing the updated API routes
  - Confirming the graph type changes work as expected

### TODO Item
- You've noted a "weird double-close bug on `RunnerInputDialog`" in your TODO list. Is this something you plan to fix before merging, or would you prefer to track it separately?

## What I Like About This PR

- Great job on reducing unnecessary re-renders in the Builder UI! The bead management improvements look solid.
- The refactoring of the run input workflow is much cleaner, using `AgentRunDetailsView` for consistency.
- The type improvements in the Graph model and API endpoints will help maintain code quality.
- Moving webhook-related code to a dedicated endpoint makes the API structure more logical.

Once you've added a test plan and addressed the feedback, this PR should be ready for another review. Let me know if you have any questions!

**Comment by AutoGPT-Agent** [comment]
Thank you for your comprehensive PR! The Builder UX improvements and credentials support look well-implemented, with a thoughtful approach to reducing unnecessary re-renders and cleaning up the run input UI components.

Before this can be merged, please:

1. Complete the test plan section in the PR checklist. Since this involves significant UI changes and credential handling, a thorough test plan is important.

2. Address the TODO item you mentioned about fixing the "weird double-close bug on RunnerInputDialog".

The changes look sound overall - I particularly like the refactoring of the bead management logic to improve performance and the cleaner approach to handling credentials. The move to direct action async callbacks instead of fire-and-forget methods in `useAgentGraph` is also a good architectural improvement.

Once you've completed the test plan and addressed the bug, this should be ready for final review and merging.

**Comment by AutoGPT-Agent** [comment]
Thank you for your PR to improve the Builder UX and add credentials support! This is a substantial contribution that will improve the platform experience.

### What Looks Good
- Your changes are well-structured and align with the PR title
- The before/after videos clearly demonstrate the improvements
- Good refactoring to reduce unnecessary re-renders and improve canvas performance
- Your code adds credentials support to the Builder in a way that's consistent with the Library

### Required Changes Before Merging

1. **Test Plan**: You need to complete the test plan in the PR description. Since your changes affect both the frontend builder UX and credentials functionality, please outline how you've tested these features. For example:
   - Testing the credentials input in Builder with various credential types
   - Verifying smooth canvas performance with large graphs
   - Testing that the run inputs are correctly passed to execution

2. **TODO Item**: You've noted a "weird double-close bug on RunnerInputDialog" - is this something you plan to fix in this PR or track separately? If it's a known issue but not blocking, please create an issue for it and reference it.

3. **Backend Changes**: Your API refactoring looks good, but please ensure these endpoint changes are reflected in any client documentation if applicable.

Once you've added the test plan and addressed the TODO item (either by fixing it or creating a tracking issue), this PR should be ready for another review. The functionality changes look solid and well-implemented.

**Comment by AutoGPT-Agent** [comment]
Thanks for your detailed PR! The changes look good overall, implementing credentials support in the Builder's Run UX and improving canvas performance.

### Feedback

1. **Incomplete checklist**: Please fill out the test plan checklist section, which is currently empty. Since this is a significant change to the Builder UX, a comprehensive test plan is important.

2. **Remaining TODO**: There's a TODO item about fixing a "weird double-close bug on `RunnerInputDialog`". Can you address this before merging, or create a follow-up issue to track it?

3. **Testing recommendations**:
   - Test with various input types in the Builder
   - Verify credentials work correctly
   - Check edge cases like empty inputs
   - Test the performance improvements with large graphs

The refactoring work and code organization improvements look solid! Nice job on making the Builder experience smoother and aligning its UX with the Library.

Please update the PR with a completed checklist and a plan for the remaining TODO before proceeding.

**Comment by AutoGPT-Agent** [comment]
Thank you for your detailed PR! The changes to improve the Builder UX and add credentials support look comprehensive and well-structured.

However, before we can approve this for merging, there are a couple of items that need attention:

1. **Missing Test Plan**: For a PR of this size and complexity, we need a complete test plan that details how you've tested these changes. Please fill out the test plan section in the PR checklist to ensure the changes have been adequately tested.

2. **Unresolved TODO**: You mentioned a "weird double-close bug on `RunnerInputDialog`" in your TODO list. This should be resolved before the PR is merged.

The code changes themselves look solid. I particularly like:
- How you've reduced unnecessary re-renders to make canvas movement smoother
- The refactoring of the credentials support to make it consistent with the Library
- The cleanup of the run input UI components

Please update the PR with a complete test plan and resolve the noted TODO item, and we'll be happy to review again.

**Comment by AutoGPT-Agent** [comment]
Thank you for this comprehensive PR that adds credentials support to the Builder UI and improves performance. The changes look well-structured and the refactoring of the frontend components should lead to better code organization and improved UX.

However, before this can be approved for merging, please complete your PR checklist:

- You've marked that you've listed your changes (which is excellent and detailed), but you haven't filled out the test plan section or indicated that you've tested according to a plan.
- For a change of this magnitude that touches multiple areas (frontend UI, backend API, etc.), it's important to have a clear test plan that verifies the functionality works as expected.

Please add a test plan and check it off to confirm you've verified your changes work properly. This should include testing:
- Basic builder functionality with the new smooth panning/zooming
- Creating and running agents with credentials
- The flow from Builder to execution with the new Runner UI implementation

Also, please address the "TODO" you've noted in the PR description about the "weird double-close bug on RunnerInputDialog" - either fix it or create an issue to track it if it's not blocking.

Once the checklist is completed with a proper test plan, this PR looks ready for another review.

File: autogpt_platform/frontend/src/lib/autogpt-server-api/types.ts [hunk 1]
```diff
@@ -24,7 +24,7 @@ export enum BlockCostType {
 export type BlockCost = {
   cost_amount: number;
   cost_type: BlockCostType;
-  cost_filter: { [key: string]: any };
+  cost_filter: Record<string, any>;
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/lib/autogpt-server-api/types.ts:27]: Apparently these aren't 100% equivalent:

```ts
type A = { [key: string]: any };
type B = Record<string, any>;

type AKey = keyof A;  // string | number
type BKey = keyof B;  // string
```

File: autogpt_platform/frontend/src/lib/autogpt-server-api/types.ts [hunk 2]
```diff
@@ -37,14 +37,12 @@ export type Block = {
   outputSchema: BlockIORootSchema;
   staticOutput: boolean;
   uiType: BlockUIType;
-  uiKey?: string;
   costs: BlockCost[];
-  hardcodedValues: { [key: string]: any } | null;
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/lib/autogpt-server-api/types.ts:42]: These aren't part of the API so I moved them to `BlocksControl`

File: autogpt_platform/frontend/src/components/CustomEdge.tsx [hunk 3]
```diff
@@ -87,89 +94,80 @@ export function CustomEdge({
...
-      for (let i = 0; i < beadUp - created; i++) {
-        newBeads.push({ t: 0, targetT: 0, startTime: Date.now() });
-      }
-
-      const b = setTargetPositions([...beads, ...newBeads]);
-      return { beads: b, created: beadUp, destroyed };
-    });
-
-    // Remove beads if not animating
-    if (visualizeBeads !== "animate") {
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/components/CustomEdge.tsx:110]: As far as I could see, `visualizeBeads` is hardcoded to `"animate"` so this was dead code.

File: autogpt_platform/frontend/src/components/Flow.tsx [hunk 4]
```diff
@@ -453,10 +451,9 @@ const FlowEditor: React.FC<{
     return uuidv4();
   }, []);
 
-  const { x, y, zoom } = useViewport();
-
   // Set the initial view port to center the canvas.
   useEffect(() => {
+    const { x, y } = getViewport();
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/components/Flow.tsx:456]: `useViewport()` causes a re-render every time the viewport shifts or zooms 💀

File: autogpt_platform/frontend/src/components/Flow.tsx [hunk 5]
```diff
@@ -656,6 +655,30 @@ const FlowEditor: React.FC<{
...
 
+  const handleRunButton = useCallback(async () => {
+    if (isRunning) return;
+    if (!savedAgent) {
+      toast({
+        title: `Please save the agent first, using the button in the left sidebar.`,
+      });
+      return;
+    }
+    await saveAgent();
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/components/Flow.tsx:661]: To make *this* work I had to rewrite most of `useAgentGraph` so it returns a promise we can wait for

File: autogpt_platform/frontend/src/lib/utils.ts [hunk 6]
```diff
@@ -227,13 +228,6 @@ export function getPrimaryCategoryColor(categories: Category[]): string {
   );
 }
 
-export function filterBlocksByType<T>(
-  blocks: T[],
-  predicate: (block: T) => boolean,
-): T[] {
-  return blocks.filter(predicate);
-}
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/lib/utils.ts:235]:
```ts
filterBlocksByType(blocks, ...) == blocks.filter(...)
```

File: autogpt_platform/frontend/src/lib/utils.ts [hunk 7]
```diff
@@ -397,6 +391,14 @@ export function isEmptyOrWhitespace(str: string | undefined | null): boolean {
   return !str || str.trim().length === 0;
 }
 
+export function isEmpty(value: any): boolean {
+  return (
+    value === undefined ||
+    value === "" ||
+    (typeof value === "object" && _isEmpty(value))
+  );
+}
```
Comment (Pwuts) [inline autogpt_platform/frontend/src/lib/utils.ts:400]: I found out the hard way that `lodash.isEmpty(1337)` returns `true`, because `1337` is not a non-empty string nor a non-empty object. This `isEmpty` implementation wraps `lodash.isEmpty` and does handle numbers correctly.

**Comment by AutoGPT-Agent** [comment]
Thanks for this PR implementing builder credentials support and improving UX! The changes look comprehensive and well-structured. 

The code looks good overall - I particularly like how you've improved the bead management logic to make canvas navigation smoother, and the refactoring of the run input UI components is a nice cleanup.

However, I noticed the checklist in your PR description is incomplete. You've marked that you've clearly listed changes (which you have done very well), but the test plan section is unchecked and empty. For a feature this important that touches both frontend and backend, we need a clear testing strategy.

Could you please:

1. Create and complete a test plan that covers your changes
2. Confirm you've tested according to that plan
3. Address the "TODO: Fix weird double-close bug on `RunnerInputDialog`" item you mentioned

This PR makes substantial improvements to the platform, but we need to ensure everything works as expected before merging, especially since it involves credential handling which has security implications.

Once these items are addressed, this PR looks ready for another review.

**Comment by AutoGPT-Agent** [comment]
Thank you for this well-structured PR! The improvements to the Builder's credential support and UX are substantial and well-implemented.

### Feedback

#### Great job on:
- Clear explanation with before/after screenshots
- Comprehensive refactoring of the run input UI
- Performance improvements for canvas rendering
- Aligning naming between frontend and backend
- Moving webhook setup route to a more logical location

#### Issues to address:

1. **Test completion needed**: You've outlined a comprehensive test plan, but the checklist items are not marked as completed. Please complete the testing according to your plan and check off those items.

2. **PR size**: This is a sizeable PR with many changes. While the changes are cohesive and related to the stated purpose, larger PRs are inherently riskier. Consider breaking future changes into smaller, more focused PRs if possible.

#### Additional considerations:

- The changes to `useAgentGraph` significantly improve the architecture, making the code more maintainable. Nice work on this refactoring.
- The move from fire-and-forget methods to direct action callbacks is a good pattern improvement.
- The rework of UI components for credentials input and run management looks much cleaner and more consistent.

Please complete the testing according to your test plan and update the PR by checking off those items in the checklist. Once the testing is complete, this PR should be ready for approval.

**Comment by AutoGPT-Agent** [comment]
This PR looks good overall! The implementation for Builder credentials support and UX improvements is well-structured, and the test plan is thorough.

There's just one small issue to fix before this can be merged:

- The parent checkbox for "I have tested my changes according to the test plan" should be checked since all the sub-items are checked. Please update this in the PR description.

Some observations about your implementation:

1. Great work on the refactoring to use `GraphMeta` and improve type safety throughout the codebase.

2. The UX improvements for moving around the canvas look excellent - reducing unnecessary re-renders should make the builder experience much smoother.

3. Good refactoring of the run input UI components to leverage the existing `AgentRunDraftView` - this creates consistency between the Builder and Library experiences.

4. The move of the webhook setup endpoint to `/library/presets/setup-trigger` makes sense from an API organization perspective.

Once you check that parent checkbox in the PR description, this should be ready to merge!

File: autogpt_platform/frontend/src/components/Flow.tsx [hunk 8]
```diff
@@ -653,39 +634,45 @@ const FlowEditor: React.FC<{
...
+        onClick: history.undo,
+      },
+      {
+        label: "Redo",
+        icon: <IconRedo2 />,
+        onClick: history.redo,
+      },
+    ],
+    [],
+  );
```
Comment (0ubbe) [inline autogpt_platform/frontend/src/components/Flow.tsx:651]: What is the benefit of `useMemo` here?  So har this looks like something  easy to render 👀

File: autogpt_platform/backend/backend/data/graph.py [hunk 9]
```diff
@@ -612,6 +619,18 @@ def from_db(
...
+class GraphMeta(Graph):
+    user_id: str
+
+    # Easy work-around to prevent exposing nodes and links in the API response
+    nodes: list[NodeModel] = Field(default=[], exclude=True)  # type: ignore
+    links: list[Link] = Field(default=[], exclude=True)
+
+    @staticmethod
+    def from_graph(graph: GraphModel) -> "GraphMeta":
+        return GraphMeta(**graph.model_dump())
```
Comment (majdyz) [inline autogpt_platform/backend/backend/data/graph.py:631]: Shouldn't we override the nodes and links here to [] too ?

File: autogpt_platform/backend/backend/data/graph.py [hunk 10]
```diff
@@ -667,13 +686,13 @@ async def get_graphs(
...
+    graph_models: list[GraphMeta] = []
     for graph in graphs:
         try:
-            graph_model = GraphModel.from_db(graph)
-            # Trigger serialization to validate that the graph is well formed.
-            graph_model.model_dump()
-            graph_models.append(graph_model)
+            graph_meta = GraphModel.from_db(graph).meta()
+            # Trigger serialization to validate that the graph is well formed
+            graph_meta.model_dump()
```
Comment (majdyz) [inline autogpt_platform/backend/backend/data/graph.py:694]: unrelated change, but how could a graph_model is loaded from the previous command but not well formed? and is it the same as model_validate?

File: autogpt_platform/frontend/src/components/CustomEdge.tsx (hunk 3 above)
Comment (majdyz) [reply autogpt_platform/frontend/src/components/CustomEdge.tsx:110]: Actually we should not animate static link or the first load of the graph,

**Review by majdyz — APPROVED** [review]
I'm excited for this!

File: autogpt_platform/frontend/src/components/runner-ui/RunnerInputUI.tsx [hunk 11]
```diff
@@ -1,87 +1,90 @@
...
+          scheduleName: string,
+          inputs: Record<string, any>,
+          credentialsInputs: Record<string, CredentialsMetaInput>,
+        ) => {
+          await doCreateSchedule(
+            cronExpression,
+            scheduleName,
+            inputs,
+            credentialsInputs,
+          );
```
Comment (0ubbe) [inline autogpt_platform/frontend/src/components/runner-ui/RunnerInputUI.tsx:65]: What happens if this fails? Shouldn't we show an error toast?

File: autogpt_platform/frontend/src/components/monitor/FlowInfo.tsx [hunk 12]
```diff
@@ -62,90 +64,30 @@ export const FlowInfo: React.FC<
       (selectedVersion == "all" ? flow.graph_version : selectedVersion),
   );
 
+  const hasInputs = Object.keys(flow.input_schema.properties).length > 0;
+  const hasCredentialsInputs =
+    Object.keys(flow.credentials_input_schema.properties).length > 0;
```
Comment (0ubbe) [inline autogpt_platform/frontend/src/components/monitor/FlowInfo.tsx:69]: Should we be more defensive when accessing `input_schema ` ? Gave us some problems recently 🙇🏽

File: autogpt_platform/frontend/src/components/edit/control/BlocksControl.tsx [hunk 13]
```diff
@@ -45,16 +73,20 @@ interface BlocksControlProps {
...
   pinBlocksPopover,
   flows,
   nodes,
-}) => {
+}: BlocksControlProps) {
   const [searchQuery, setSearchQuery] = useState("");
+  const deferredSearchQuery = useDeferredValue(searchQuery);
   const [selectedCategory, setSelectedCategory] = useState<string | null>(null);
 
+  // Preprocess blocks with cached string operations
```
Comment (0ubbe) [inline autogpt_platform/frontend/src/components/edit/control/BlocksControl.tsx]: The hook already has a comment, so this comment may not be needed?

**Review by 0ubbe — DISMISSED** [review]
Looking good overall, some minor comments 👍🏽 ✅

File: autogpt_platform/backend/backend/data/graph.py (hunk 9 above)
Comment (Pwuts) [reply autogpt_platform/backend/backend/data/graph.py:631]: No, that would break `input_schema` and `credentials_input_schema`. But `GraphMeta` omits them from serialization so from the response. :)

File: autogpt_platform/backend/backend/data/graph.py (hunk 10 above)
Comment (Pwuts) [reply autogpt_platform/backend/backend/data/graph.py:694]: On serialization, the computed fields are rendered, which can fail if e.g. the graph contains a node with an unknown `block_id`. By calling `model_dump` here, we can gracefully handle such cases.

File: autogpt_platform/frontend/src/components/monitor/FlowInfo.tsx (hunk 12 above)
Comment (Pwuts) [reply autogpt_platform/frontend/src/components/monitor/FlowInfo.tsx:69]: In my opinion: no. The frontend types should match the backend behavior, and then the type checker will tell us when we're not defensive enough. I have checked the frontend types against the backend implementation, updated a few things, and they should match now.

According to the type definition (based on the backend behavior), `flow.input_schema` is always present and always has a `properties` attribute. Adding `?` without understanding why it's breaking or updating the type is just asking for more bugs to happen.

File: autogpt_platform/frontend/src/components/Flow.tsx (hunk 8 above)
Comment (Pwuts) [reply autogpt_platform/frontend/src/components/Flow.tsx:651]: Reducing unnecessary re-renders of `ControlPanel`, which it is passed to.

File: autogpt_platform/frontend/src/components/runner-ui/RunnerInputUI.tsx (hunk 11 above)
Comment (Pwuts) [reply autogpt_platform/frontend/src/components/runner-ui/RunnerInputUI.tsx:65]: `useAgentGraph.createRunSchedule(..)` does this: https://github.com/Significant-Gravitas/AutoGPT/pull/10323/files/a0c5169b982c158b4dd3b59f96a793ce09eb3d8a#diff-0ad1917a409dfebde02126122593f0fccab47c6ba44c7fb96336f69eceffbd70L1104

File: autogpt_platform/frontend/src/components/edit/control/BlocksControl.tsx (hunk 13 above)
Comment (Pwuts) [reply autogpt_platform/frontend/src/components/edit/control/BlocksControl.tsx]:
```suggestion
```

File: autogpt_platform/frontend/src/components/CustomEdge.tsx (hunk 3 above)
Comment (Pwuts) [reply autogpt_platform/frontend/src/components/CustomEdge.tsx:110]: Is that something I need to fix in this PR or a note to fix sometime in the future? Afaik always-on animation is the current behavior.

File: autogpt_platform/frontend/src/components/monitor/FlowInfo.tsx (hunk 12 above)
Comment (0ubbe) [reply autogpt_platform/frontend/src/components/monitor/FlowInfo.tsx:69]: Agree, if we can trust the type then so be it 🙇🏽

File: autogpt_platform/frontend/src/components/type-based-input.tsx
Comment (Pwuts) [reply autogpt_platform/frontend/src/components/type-based-input.tsx:1]: reverted

**Review by majdyz — APPROVED** [review]
